app/services/cocina_generator/structural/generator.rb

```diff
@@ -24,38 +26,63 @@ def generate
       end
 
       # 1) Start with the existing filesets from SDR
-      # 1) Only keep the filesets that have attached_files that are preserved
-      # 2) Add new filesets that have attached_files in the local store.
+      # 2) Rebuild filesets that have attached_files that are preserved
+      # 3) Add new filesets that have attached_files in the local store.
       def build_filesets
         filesets = find_preserved_filesets
-        filesets + work_version.staged_files.map.with_index(filesets.size + 1) { |af, n| build_fileset(af, n) }
+        # rebuild the the ones that are still preserved
+        filesets.each do |filename, fileset|
+          rebuilt_fileset = build_fileset(attached_file: get_preserved_file(filename), fileset:)
+          filesets[filename] = rebuilt_fileset
+        end
+        filesets.values + work_version.staged_files.map { |af| build_fileset(attached_file: af) }
       end
 
-      # @return [Array<Cocina::FileSet>] the list of items where all files are already preserved.
+      # @return [Hash] hash with key is filename, value is fileset cocina for preserved files.
       def find_preserved_filesets
-        return [] unless cocina_obj
+        return {} unless cocina_obj
+
+        preserved_file_names = work_version.preserved_files.map(&:filename).map(&:to_s)
+        create_filesmap(preserved_file_names)
+      end
+
+      def get_preserved_file(filename)
+        work_version.preserved_files.find { |attached_file| attached_file.filename.to_s == filename }
+      end
 
-        cocina_obj.structural.contains.select do |fileset|
-          existing_file_names = Set.new(fileset.structural.contains.map(&:filename))
-          preserved_file_names = Set.new(work_version.preserved_files.map(&:filename).map(&:to_s))
-          preserved_file_names.superset?(existing_file_names)
+      def create_filesmap(preserved_file_names)
+        files_map = {}
+        cocina_obj.structural.contains.to_a.each do |fileset|
```

Comment:
Is the `#to_a` being used here to handle a `nil` contains? (Because `contains` should already be an array, right?)

Reply:
Right, `contains` should be an array. I had been getting some errors about not being able to use `.each` on a cocina object, and that seemed to address it. But if I remove the `to_a` now, I'm no longer getting that error. So perhaps it was something to do an earlier version I have been working on. At any rate, I am taking this out and will test.